modules/utils.py

```diff
@@ -0,0 +1,109 @@
+import torch
+from torch.utils.data import TensorDataset, Subset
+import numpy as np
+import matplotlib.pyplot as plt
+
+# --- EVALUATION ----
+
+def _accuracy(preds, labels):
+    return (preds == labels).float().mean().item()
+
+def evaluate(model, test_dataloader):
+    model.eval()
+
+    loss_arr = []
+    accuracy_arr = []
+
+    device = "cuda"
+
+    for batch in test_dataloader:
+        batch = tuple(t.to(device) for t in batch)
+
+        with torch.no_grad():
+            inputs = {'input_ids':      batch[0],
+                      'attention_mask': batch[1],
+                      'labels':         batch[2]}
+
+            outputs = model(**inputs)
+            loss, logits = outputs[:2]
+
+            preds = torch.argmax(logits, dim=1)
+            labels = inputs['labels']
+
+            loss_arr.append(loss.item())
+            accuracy_arr.append(_accuracy(preds, labels))
+
+    model.train()
+    avg_loss = sum(loss_arr) / len(loss_arr)
+    avg_accuracy = sum(accuracy_arr) / len(accuracy_arr)
+    return avg_loss, avg_accuracy
+
+# --- PARTITIONING ---
+
+def partition_dataset_homogeneous(dataset: TensorDataset, num_clients=3):
+    indices = np.random.permutation(len(dataset))
+    splits = np.array_split(indices, num_clients)
+    return [Subset(dataset, split) for split in splits]
+
+def partition_dataset_heterogeneous(dataset: TensorDataset, num_clients=3):
```

Comment:
What's the point of this parameter if proportions have only three elements anyway? Maybe pass proportions and do `for client in range(len(proportions))`?

Honestly, just being annoying...